test/src/containerTestBase.ts

```diff
@@ -0,0 +1,139 @@
+				return true;
+			}
+	
+			// Build or pull image specified by devcontainer.json
+			it('should start', async function () {
+				this.slow(120000);
+				this.timeout(0);
+	
+				devContainer = await testContainerResolver(definitionPath);
+				assert.equal(await fetchAgentCommit(devContainer.resolvedAuthority), devContainer.params.product.commit);
+	
+				return true;
+			});
+	
+			it('should execute', async function () {
+				this.slow(120000);
+				this.timeout(0);
+	
+				const result = await execTestScript(devContainer, path.resolve(rootFolder, '..'));
+				assert.isNull(result.error, result.error + ': ' + result.stderr);
+	
+				return true;
+			});
+	
+			it('should clean up', async function () {
+				this.slow(120000);
+				this.timeout(0);
+	
+				const result = await cleanUpTest(devContainer);
+				assert.isNull(result.error, result.error + ': ' + result.stderr);
+	
+				return true;
+			});
+	
+			return true;
+		});
+	});
+}
+
+
+export function validateDevContainerJson(definitionPath: string, schema: any) {
+	log('debug', `Validating devcontainer.json at ${definitionPath}`);
+	const jsonPath = path.join(definitionPath, '.devcontainer', 'devcontainer.json');
+	const json = fs.existsSync(jsonPath) ? jsonc.readSync(jsonPath) :
+		jsonc.readSync(path.join(definitionPath, '.devcontainer.json'));
+
+	return tv4.validate(json, schema);
+}
+
+export async function testContainerResolver(definitionPath: string): Promise<ResolverResult> {
+
+	log('debug', `Finding and reading devcontainer.json at ${definitionPath}`);
+	const configPath = await getDevContainerConfigPathIn(definitionPath);
+	const config: any = configPath && await readDevContainerConfigFile(configPath);
+
+	const params = await createTestParams();
+	params.getExtensionsToInstall = () => config.extensions;
+	params.cwd = definitionPath;
+	params.progress = {
+		report: (value: any) => { log('trace', value.message); }
+	};
+	params.output = {
+		write: (message) => { log('trace', message); }
+	};
+
+	log('debug', 'Running resolver.');
+	return await resolve(params, config);
+}
+
+export async function execTestScript(devContainer: ResolverResult, vscodeDevConPath: string = '') {
+	const container = await devContainer.params.docker.getContainer(devContainer.properties.id);
+	const containerInfo = await container.inspect();
+	const workingDir = findMountDestination(containerInfo, devContainer.params.cwd, vscodeDevConPath);
+	const result = await exec(`docker exec ${containerInfo.Id} /bin/sh -c "cd ${workingDir} && if [ -f test-project/test.sh ]; then cd test-project && chmod +x test.sh && ./test.sh; else ls -a; fi"`)
+	log('trace', result.stdout);
+	if (result.error) {
+		log('trace', result.error + ': ' + result.stderr);
+	}
+
+	return result;
+}
+
+export async function cleanUpTest(devContainer: ResolverResult) {
+
+	log('debug', 'Stopping container');
+	devContainer.params.shutdowns.forEach(async (shutdown: (rebuild?: boolean) => {}) => await shutdown(false));
```

Comment:
Is the goal here to await for all of them to shutdown? If so, I don't think awaiting shutdown inside that forEach is what you'd want.

Something like this might have that behaviour though:

<pre>
const shutdownAllContainers = devContainer.params.shutdowns.map(async (shutdown: async (rebuild?: boolean) => Promise<void>) =>  shutdown(false));
await Promise.all(shutdownAllContainers);
</pre>